File: fuargs/topdir.py
import os.path, sys

class TopDir:
    def __init__(self, marker_file):
        self.marker_file = marker_file

    def topdir(self, pn):
        prev_pn = None
        while 1:
            if prev_pn is not None and pn == prev_pn: # at the top
                raise Exception("topdir: no marker file found: %s" % self.marker_file)

            marker_pn = os.path.join(pn, self.marker_file)
            if os.path.exists(marker_pn):
                if not os.path.isfile(marker_pn):
                    raise Exception("topdir: name clash, marker file %s is directory or else" % marker_pn)
                return pn
            prev_pn = pn
            pn = os.path.dirname(pn)

def topdir(marker_file = 'TOP-DIR'):
    o = TopDir(marker_file)
    argv0 = sys.argv[0]
    pn = os.path.realpath(os.path.dirname(argv0))

    return o.topdir(pn)

if __name__ == "__main__":
    print("__file__:", __file__)
    print(topdir(marker_file = "TOP-DIR"))

File: cpp/src/topdir.cpp
#include <stdexcept>
using namespace std;

#include <experimental/filesystem>
namespace fs = std::experimental::filesystem;

#include <kvan/topdir.h>

// see also https://unix.stackexchange.com/a/317215/37162
//
static fs::path get_self_path()
{
  fs::path self_exe = "/proc/self/exe";
  return fs::read_symlink(self_exe);
}

TopDirImpl::TopDirImpl(const char* marker_file)
{
  fs::path argv0 = get_self_path();
  if (!argv0.has_filename()) {
    throw runtime_error(__func__);
  }
  
  fs::path c_dir = argv0.remove_filename();
  bool marker_file_found = false;
  do {
    fs::path marker_pn = c_dir / marker_file;
    if (fs::exists(marker_pn)) {
      topdir = marker_pn.remove_filename();
      marker_file_found = true;
      break;
    }
    c_dir = fs::canonical(c_dir / "..");
  } while (c_dir != fs::path("/") && fs::exists(c_dir));

  if (!marker_file_found) {
    throw runtime_error("TopDir::init: marker file not found");
  }    
}
  
